### PR TITLE
[speculator training] Speculator training

### PR DESCRIPTION
**Pull request by daviswer**
Add support for speculator training, piggybacking off the existing training utilities. 

Training script and speculator-specific utilities are inside the new `speculator` subfolder. 

Uses distributed setup, checkpointing, and dataloaders from this repo. Adds speculator-specific fields to the training config file (to be ignored during non-speculator training). It might make more sense to pull these new fields out into a separate config subclass under speculator utilities - open to suggestions. 

Uses speculator architecture from fms-extras. 

Uses altered Llama-7b and `generate()` function from base fms, allowing the speculator to access embedding vectors, not just logits/token predictions. ~~Do not merge this until [that](https://github.com/foundation-model-stack/foundation-model-stack/pull/193) issue can be resolved.~~

### CODE REVIEW

**Comment by daviswer** [comment]
Plan is to move the `include_embeds=True` versions of Llama/GPTBigCode/generate() into `fms-extras`. Once that is done I'll update the relevant imports here and then we can push this in

**Comment by daviswer** [comment]
I've pulled all the `include_embeds` stuff out of `fms` into here. We now have `EmbedLLaMA` and `EmbedGPTBigCode` subclasses that override the corresponding forward function, and an altered version of `generate` for use only with this script. We register the subclassed models for use with `get_model` in the training script.

**Comment by daviswer** [comment]
Code is ready for review - mypy errors are import errors, it doesn't have `fms-extras` and it doesn't like the local import of `train_speculator_utils`. Should I move the speculator subfolder under `fms_fsdp` so that I can use an absolute path for that?

File: fms_fsdp/config/training.py [hunk 1]
```diff
@@ -42,3 +42,12 @@ class train_config:
 
     # compile
     use_torch_compile: bool = False
+
+    # speculator training
+    model_path: str = "/lustre/llama_weights/hf/7B-F/"
```
Comment (JRosenkranz) [inline fms_fsdp/config/training.py:47]: should this be some other path for a default?

File: fms_fsdp/utils/dataloader_utils.py [hunk 2]
```diff
@@ -1,16 +1,43 @@
...
     Buffer_Dataset,
     Preload_Buffer_Dataset,
     Preprocess_Dataset,
     Sampling_Dataset,
     Scalable_Shard_Dataset,
-    Streaming_Doc_Dataset,
 )
 
 
+def parse_data_args(datas, weights):
```
Comment (JRosenkranz) [inline fms_fsdp/utils/dataloader_utils.py:12]: type hints needed

File: fms_fsdp/utils/dataloader_utils.py [hunk 3]
```diff
@@ -1,16 +1,43 @@
...
+            return [x]
+        else:
+            raise ValueError(f"arg input {x} cannot be parsed.")
+
+    datas = splitstrip(datas)
+    weights = [float(x) for x in splitstrip(weights)]
+    return datas, weights
+
+
+def causal_lm(data_seq, prompt_len=1):
```
Comment (JRosenkranz) [inline fms_fsdp/utils/dataloader_utils.py:29]: type hints needed

File: fms_fsdp/utils/dataloader_utils.py [hunk 4]
```diff
@@ -86,30 +105,15 @@ def causal_lm(data_seq, prompt_len=1):
...
+        cfg.seq_length if causal_lm not in postprocess else cfg.seq_length + 1,
         drop_final_token=cfg.sep_token,
         pack_hard=True,
     )
     # Shuffle outputs in length 10k buffer. Consecutive lines appear 10k steps apart on average.
     data = Preload_Buffer_Dataset(data, 10000)
-    # Split line into input and target for the CLM task.
-    data = Preprocess_Dataset(data, causal_lm)
+    # Apply desired postprocessing steps in sequence
+    data = Preprocess_Dataset(data, torch.IntTensor)
```
Comment (JRosenkranz) [inline fms_fsdp/utils/dataloader_utils.py:115]: is this just wrapping with IntTensor?

File: speculator/train_speculator.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    speculator, optimizer, train_loader, start_step, tokens_seen = checkpointer.load(
+        speculator,
+        optimizer,
+        train_loader,
+        path=os.path.join(cfg.ckpt_load_path, "checkpoints/"),
+    )
+
+    # LR schedule
+    warmup_interval1 = min(2000, cfg.stage2_start_step // 20)
+    stage1_schedule = lambda x: min(
```
Comment (JRosenkranz) [inline speculator/train_speculator.py:165]: can we add some comments around this

File: speculator/train_speculator.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        0.1
+        + 0.5
+        * (1 - 0.1)
+        * (
+            1
+            + math.cos(min(x, cfg.stage2_start_step) / cfg.stage2_start_step * math.pi)
+        ),
+    )
+    warmup_interval2 = min(2000, (cfg.num_steps - cfg.stage2_start_step) // 20)
+    stage2_schedule = lambda x: min(
```
Comment (JRosenkranz) [inline speculator/train_speculator.py:176]: can we add some comments around this

File: speculator/train_speculator_utils.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    if not batched:
+        result = result[0]
+
+    if include_embeds:
+        return result, embeds
+
+    return result
+
+
+def stage1_loss(model, speculator, input, loss_fn, ddp_stats):
```
Comment (JRosenkranz) [inline speculator/train_speculator_utils.py:107]: Could you add docstrings for this

File: speculator/train_speculator_utils.py [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+    for i in range(preds.size(0)):
+        targ = input[:, i + 2 : preds.size(2) + i + 2]  # b n
+        loss = loss_fn(preds[i].reshape(-1, preds.size(3)), targ.long().reshape(-1))
+        losses.append(loss)
+        ddp_stats[2 + i] += loss.item()
+    loss = sum(losses)
+    return loss, ddp_stats, input.numel()
+
+
+def stage2_loss(cfg, model, speculator, input, loss_fn, ddp_stats):
```
Comment (JRosenkranz) [inline speculator/train_speculator_utils.py:126]: Docstrings for this as well

File: speculator/train_speculator_utils.py [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+    for i in range(preds.size(0)):
+        targ = targs[:, i + 1 : preds.size(2) + i + 1]  # b n
+        loss = loss_fn(preds[i].reshape(-1, preds.size(3)), targ.long().reshape(-1))
+        losses.append(loss)
+        ddp_stats[2 + i] += loss.item()
+    loss = sum(losses)
+    return loss, ddp_stats, targs.numel()
+
+
+def train_speculator(
```
Comment (JRosenkranz) [inline speculator/train_speculator_utils.py:158]: docstrings and type hints

**Review by JRosenkranz — COMMENTED** [review]
some initial comments on this PR

File: fms_fsdp/utils/dataloader_utils.py (hunk 4 above)
Comment (daviswer) [reply fms_fsdp/utils/dataloader_utils.py:115]: Yes - turn list outputs into torch tensors before applying any user-specified preprocess functions

**Comment by AlpinDale** [comment]
Hi! What's the status on this PR? I'd like to train a few speculator models, but I'm not sure how to get started, due to a lack of documentation...

**Comment by JRosenkranz** [comment]
> Hi! What's the status on this PR? I'd like to train a few speculator models, but I'm not sure how to get started, due to a lack of documentation...

Hi @AlpinDale Working on getting the documentation and code ready for this. Planning to have sometime in the next 3 weeks. Will keep you updated if we get this sometime sooner.

**Comment by AlpinDale** [comment]
Thanks for the reply, @JRosenkranz 

I'd love to wait but I have access to a large cluster of H100s for a limited time, so I wanted to make the most out of it by training as many MLPSpeculator models as possible, on various popular models. If its doable, I'd love some basic instructions on how to get this PR running and start train runs; I can figure out the rest. Different story if the PR itself isn't ready, however 😅

**Comment by vdabravolski** [comment]
Hi, adding +1 to @AlpinDale. We are interested to experiment with MLP speculator, specifically, on latest Llama3.1 models. 

Excellent work overall @JRosenkranz !

**Comment by sahilsuneja1** [comment]
Hi @AlpinDale @vdabravolski,

PR35 is outdated. We expect to release a stable code version in about 3 weeks.

We understand @AlpinDale's urgency and are trying to put this PR in shape so that you can use it in the interim. We hit issues running it against the main branches of foundation-model-stack and fms-extras, and are working on resolving it. If that doesn't work out we can point you to the specific branches for foundation-model-stack, fms-extras and fms-fsdp repos in the meanwhile so that you can train custom speculators, while we work on polishing them and merging them into their respective mains. 

There are already a bunch of speculators available [here](https://huggingface.co/collections/ibm-fms/speculators-66a1b6838f0d2327e0a3a8c3) and [here](https://huggingface.co/collections/ibm-granite/granite-speculators-664b97a44ddc5640e8cd73ac), in case there is any overlap with your requirements. For example, the llama3-70b speculator works for llama3.1-70b as well as mentioned [here](https://huggingface.co/ibm-fms/llama3-70b-accelerator) (and so llama3-8b might also work for llama3.1-8b) .

**Comment by sahilsuneja1** [comment]
@AlpinDale @vdabravolski 
Done with updating PR35-- it should now work with foundation-model-stack and fms-extras main branches.
Added a sample training script containing example arguments to pass to the speculator training routine.
Most arg names should be straightforward. For more details please refer: https://arxiv.org/pdf/2404.19124 and https://pytorch.org/blog/hitchhikers-guide-speculative-decoding/

**Comment by philschmid** [comment]
Is this expected to be merged soon?

**Comment by JRosenkranz** [comment]
> Is this expected to be merged soon?

@philschmid We are expecting to have speculator training merged sometime in next 2 weeks.

**Comment by JRosenkranz** [comment]
@philschmid This has been finished and merged in #114. @philschmid The speculator training implementation is now available in main. Please let us know if you have any feedback or questions.

**Comment by JRosenkranz** [comment]
Closing in favor of #114
